src/snowflake/connector/query_context_cache.py

```diff
@@ -0,0 +1,271 @@
+#!/usr/bin/env python
+#
+# Copyright (c) 2012-2021 Snowflake Computing Inc. All rights reserved.
+#
+from __future__ import annotations
+
+from base64 import b64decode, b64encode
+from functools import total_ordering
+from hashlib import md5
+from io import BytesIO
+from logging import DEBUG, getLogger
+from threading import Lock
+
+from sortedcontainers import SortedSet
+
+from .options import pyarrow as pa
+
+logger = getLogger(__name__)
+
+
+@total_ordering
+class QueryContextElement:
+    def __init__(self, id: int, read_timestamp: int, priority: int, context: bytearray):
+        self._id = id
+        self._read_timestamp = read_timestamp
+        # priority values are 0..N with 0 being the highest priority
+        self._priority = priority
+        self._context = context
+
+    @property
+    def id(self) -> int:
+        return self._id
+
+    @property
+    def read_timestamp(self) -> int:
+        return self._read_timestamp
+
+    @read_timestamp.setter
+    def read_timestamp(self, timestamp: int) -> None:
+        self._read_timestamp = timestamp
+
+    @property
+    def priority(self) -> int:
+        return self._priority
+
+    @property
+    def context(self) -> bytearray:
+        return self._context
+
+    @context.setter
+    def context(self, ctx: bytearray) -> None:
+        self._context = ctx
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, QueryContextElement):
+            return False
+        return (
+            self._id == other.id
+            and self._read_timestamp == other.read_timestamp
+            and self._priority == other.priority
+            and self._context == other.context
+        )
+
+    def __lt__(self, other: object) -> bool:
+        if not isinstance(other, QueryContextElement):
+            raise NotImplementedError(
+                f"cannot compare QueryContextElement with object of type {type(other)}"
+            )
+        return self._priority < other.priority
+
+    def __hash__(self) -> int:
+        _hash = 31
+
+        _hash = _hash * 31 + self._id
+        _hash += (_hash * 31) + self._read_timestamp
+        _hash += (_hash * 31) + self._priority
+        if self._context:
+            _hash += (_hash * 31) + int.from_bytes(md5(self._context).digest(), "big")
+        else:
+            _hash += _hash * 31
+        return _hash
+
+    def __str__(self) -> str:
+        return f"({self._id}, {self._read_timestamp}, {self._priority})"
+
+
+class QueryContextCache:
+    QUERY_CONTEXT_SCHEMA = pa.schema(
```

Comment:
Today we install `pyarrow` as part of `.pandas`, which means `pyarrow` is not required dependency. 

We probably should have a check on the availability of `pyarrow`, and only enable `query_context_cache` when it is available. Otherwise if `pyarrow` is not installed, query will fail here.

Reply:
thanks. this was causing multiple test failures earlier